### ISSUE TITLE
ASSERT failure when nginx is starting up

### ISSUE THREAD

**Issue by bleve** [slact/nchan#683]
```
2024/02/16 10:47:53 [alert] 23911#0: worker process 24450 exited on signal 6 (core dumped)
ker process: /builddir/build/BUILD/nchan-4461dbe99aecb51bf1afe6d00404c610f6ef706e/src/store/redis/redis_nodeset.c:1729: node_connect: Assertion `node->state <= REDIS_NODE_DISCONNECTED' failed.
```

This happens with 1.3.6 and also latest git version.

So during nginx startup, when redis connection is being initialized, worker processes crash. When redis session is initialized, crashes  stop and nchan starts to work.

Sometimes there is 0 crashes, sometimes 40.

**Comment by bleve** [comment]
Also tested removing assert, result is segfault.

**Comment by slact** [comment]
What's the Redis version?

Set the core error log level to info. What's in the logfile before the first crash?

**Comment by bleve** [comment]
redis 7.2.4.

Which log level do you mean?

**Comment by slact** [comment]
in the root-level config (not inside an `http` { } block or any other block, add:

`error_log /path/to/error_logfile info;`

**Comment by bleve** [comment]
This is everything from single startup. So nchan causes some core dumps, but after it gets redis connection up it starts to work.

```
2024/02/23 14:11:51 [info] 86391#0: Using 116KiB of shared memory for nchan in /etc/nginx/nginx.conf:53
2024/02/23 14:11:51 [info] 86391#0: Using 131072KiB of shared memory for nchan in /etc/nginx/nginx.conf:53
ker process: /builddir/build/BUILD/nchan-4461dbe99aecb51bf1afe6d00404c610f6ef706e/src/store/redis/redis_nodeset.c:1729: node_connect: Assertion `node->state <= REDIS_NODE_DISCONNECTED' failed.
ker process: /builddir/build/BUILD/nchan-4461dbe99aecb51bf1afe6d00404c610f6ef706e/src/store/redis/redis_nodeset.c:1729: node_connect: Assertion `node->state <= REDIS_NODE_DISCONNECTED' failed.
ker process: /builddir/build/BUILD/nchan-4461dbe99aecb51bf1afe6d00404c610f6ef706e/src/store/redis/redis_nodeset.c:1729: node_connect: Assertion `node->state <= REDIS_NODE_DISCONNECTED' failed.
2024/02/23 14:11:52 [alert] 86393#0: worker process 86396 exited on signal 6 (core dumped)
2024/02/23 14:11:52 [alert] 86393#0: worker process 86397 exited on signal 6 (core dumped)
ker process: /builddir/build/BUILD/nchan-4461dbe99aecb51bf1afe6d00404c610f6ef706e/src/store/redis/redis_nodeset.c:1729: node_connect: Assertion `node->state <= REDIS_NODE_DISCONNECTED' failed.
2024/02/23 14:11:52 [alert] 86393#0: worker process 86395 exited on signal 6 (core dumped)
2024/02/23 14:11:52 [alert] 86393#0: worker process 86416 exited on signal 6 (core dumped)
ker process: /builddir/build/BUILD/nchan-4461dbe99aecb51bf1afe6d00404c610f6ef706e/src/store/redis/redis_nodeset.c:1729: node_connect: Assertion `node->state <= REDIS_NODE_DISCONNECTED' failed.
2024/02/23 14:11:52 [alert] 86393#0: worker process 86426 exited on signal 6 (core dumped)
ker process: /builddir/build/BUILD/nchan-4461dbe99aecb51bf1afe6d00404c610f6ef706e/src/store/redis/redis_nodeset.c:1729: node_connect: Assertion `node->state <= REDIS_NODE_DISCONNECTED' failed.
2024/02/23 14:11:52 [alert] 86393#0: worker process 86433 exited on signal 6 (core dumped)
ker process: /builddir/build/BUILD/nchan-4461dbe99aecb51bf1afe6d00404c610f6ef706e/src/store/redis/redis_nodeset.c:1729: node_connect: Assertion `node->state <= REDIS_NODE_DISCONNECTED' failed.
2024/02/23 14:11:52 [alert] 86393#0: worker process 86440 exited on signal 6 (core dumped)
ker process: /builddir/build/BUILD/nchan-4461dbe99aecb51bf1afe6d00404c610f6ef706e/src/store/redis/redis_nodeset.c:1729: node_connect: Assertion `node->state <= REDIS_NODE_DISCONNECTED' failed.
2024/02/23 14:11:52 [alert] 86393#0: worker process 86448 exited on signal 6 (core dumped)
```

**Comment by bleve** [comment]
This is nginx-1.24.0

**Comment by slact** [comment]
That log still looks incomplete. Can you show me your Nginx config? Edit out whatever you need for privacy.

**Comment by bleve** [comment]
Log is complete. And those lines actually start with `ker process:`

/etc/nginx/nginx.conf:

```
# For more information on configuration, see:
#   * Official English Documentation: http://nginx.org/en/docs/
#   * Official Russian Documentation: http://nginx.org/ru/docs/

user nginx;
worker_processes auto;
error_log logs/error.log;
pid /run/nginx/nginx.pid;

# Load dynamic modules. See /usr/share/doc/nginx/README.dynamic.
include /usr/share/nginx/modules/*.conf;

# Include events config files.
events {
    include     /etc/nginx/events.d/*.conf;
}

http {
    log_format  main    '$remote_addr - $remote_user [$time_local] "$request" '
                        '$status $body_bytes_sent "$http_referer" '
                        '"$http_user_agent" "$http_x_forwarded_for"';
    access_log  logs/access.log  main;

    sendfile            on;
    tcp_nopush          on;
    types_hash_max_size 4096;
    include             /etc/nginx/mime.types;

    # Enable session cache
    ssl_session_cache   shared:SSL:30m;

    # Default to UTF-8 on all hosts
    charset             utf-8;

    # Load modular configuration files from the /etc/nginx/conf.d directory.
    include             /etc/nginx/conf.d/*.conf;
    include             /etc/nginx/sites-enabled/*;

    # Add "Test Page" server.
    server {
        listen          80;
        listen          [::]:80;
        server_name     _;
        root            /usr/share/nginx/html;
        error_page      404  /404.html;
        error_page      500 502 503 504  /50x.html;
    }
}
```

I left out log configuration which just defines json formatted logging.

This is /etc/nginx/conf.d/pubsub.conf

```
variables_hash_max_size 4096;

upstream redis {
    nchan_redis_server redis://172.27.32.228:6379/1;
    nchan_redis_server redis://172.27.32.229:6379/1;
}

server {
    listen 443 http2 ssl;

    ssl_certificate     /etc/pki/tls/certs/star.example.com.chained.crt;
    ssl_certificate_key /etc/pki/tls/private/star.example.com.key;
    ssl_dhparam         /etc/pki/tls/private/nginx-dhparam.pem;

    server_name       pubsub-01.example.com;
    access_log        logs/pubsub_access.log json;
    error_log         logs/pubsub_error.log;

    location ~ ^/subscribe/(\w+)$ {
      nchan_subscriber;
      nchan_channel_id $1;
      nchan_redis_pass redis;
    }

    location ~ ^/publish/(\w+)$ {
      nchan_publisher;
      nchan_channel_id $1;
      nchan_redis_pass redis;
    }

    location / {
        return 404 "pubsub only";
    }
}
```

**Comment by slact** [comment]
Alright, set the loglevel to `notice` and try again please.
```
error_log logs/error.log notice;
```

**Comment by bleve** [comment]
[nchan-error.log](https://github.com/slact/nchan/files/14426586/nchan-error.log)

**Comment by slact** [comment]
Thanks. now we're getting somewhere. What's the output of the `INFO` command on 172.27.32.228:6379 ?

**Comment by slact** [comment]
Disregard the above for now. Remove one of the server addresses from the `upstreams` block and try again.

**Comment by slact** [comment]
Got it, will have a fix shortly.

**Comment by slact** [comment]
alright, this should be fixed in master with commits bcb823029c122c471d1e1f541ad58fcc703e171c and 3995f5d07d26674f9836cb894cfd651333a7869b

**Comment by bleve** [comment]
Verified. fixed.
